src/nvim/search.c

```diff
@@ -5061,3 +5061,531 @@ bool search_was_last_used(void)
+      (void)mb_ptr2char_adv((const char_u **const)&fuzpat);
+    }
+    (void)mb_ptr2char_adv((const char_u **const)&str);
+  }
+
+  // Determine if full fuzpat was matched
+  matched = *fuzpat == NUL ? true : false;
+
+  // Calculate score
+  if (matched) {
+    *outScore = fuzzy_match_compute_score(strBegin, strLen, matches,
+                                          nextMatch);
+  }
+
+  // Return best result
+  if (recursiveMatch && (!matched || bestRecursiveScore > *outScore)) {
+    // Recursive score is better than "this"
+    memcpy(matches, bestRecursiveMatches, maxMatches * sizeof(matches[0]));
+    *outScore = bestRecursiveScore;
+    return true;
+  } else if (matched) {
+    return true;  // "this" score is better than recursive
+  }
+
+  return false;  // no match
+}
+
+// fuzzy_match()
+//
+// Performs exhaustive search via recursion to find all possible matches and
+// match with highest score.
+// Scores values have no intrinsic meaning.  Possible score range is not
+// normalized and varies with pattern.
+// Recursion is limited internally (default=10) to prevent degenerate cases
+// (fuzpat="aaaaaa" str="aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa").
+// Uses char_u for match indices. Therefore patterns are limited to MAXMATCHES
+// characters.
+//
+// Returns true if 'fuzpat' matches 'str'. Also returns the match score in
+// 'outScore' and the matching character positions in 'matches'.
+static int fuzzy_match(
+    char *str,
+    char *fuzpat,
+    int *outScore,
+    matchidx_T *matches,
+    int maxMatches)
+{
+  int recursionCount = 0;
+  int len = utf_char2len(*str);
+
+  *outScore = 0;
+
+  return fuzzy_match_recursive(
+      fuzpat, str, 0, outScore, str, len, NULL,
+      matches, maxMatches, 0, &recursionCount);
+}
+
+// Sort the fuzzy matches in the descending order of the match score.
+static int fuzzy_item_compare(const void *s1, const void *s2)
+{
+  int v1 = ((fuzzyItem_T *)s1)->score;
+  int v2 = ((fuzzyItem_T *)s2)->score;
+
+  return v1 == v2 ? 0 : v1 > v2 ? -1 : 1;
+}
+
+// Fuzzy search the string 'str' in a list of 'items' and return the matching
+// strings in 'fmatchlist'.
+// If 'items' is a list of strings, then search for 'str' in the list.
+// If 'items' is a list of dicts, then either use 'key' to lookup the string
+// for each item or use 'item_cb' Funcref function to get the string.
+// If 'retmatchpos' is true, then return a list of positions where 'str'
+// matches for each item.
+static void match_fuzzy(
+    list_T *items,
+    char *str,
+    char *key,
+    Callback *item_cb,
+    int retmatchpos,
+    list_T *fmatchlist)
+{
+  long len;
+  fuzzyItem_T *ptrs;
+  long i = 0;
+  bool found_match = false;
+  matchidx_T matches[MAXMATCHES];
+
+  len = tv_list_len(items);
+  if (len == 0) {
+    return;
+  }
+
+  ptrs = xcalloc(sizeof(fuzzyItem_T), len);
+  if (ptrs == NULL) {
+    return;
+  }
+
+  // For all the string items in items, get the fuzzy matching score
+  TV_LIST_ITER(items, li, {
```

Comment:
This does not seem to work and I really don't know why.

Reply:
Did you check https://github.com/neovim/neovim/wiki/List-management-in-Neovim?

Reply:
Yes I used it to port this patch.

Reply:
@janlazo coming to this after a while, it is still a complete mistery  for me why this loop is not running at all.

Reply:
Do you have a test-case where it's not running? I've ran the PR and with the 2 other comments it seemed to run fine for simple cases.